### PR TITLE
* feat(factory.py): add support for merging consecutive messages of one role when separated with empty message of another role

### PR DESCRIPTION
**Pull request by nkvch**
Related to #3474

### CODE REVIEW

**Comment by ghost** [comment]
Hi @nkvch thanks for the issue + PR 

Can you add a unit test for this here - https://github.com/BerriAI/litellm/blob/c83348b7a85e59f9241e35aa5fff3b95a6c14433/litellm/tests/test_completion.py#L119

And confirm your PR works for all the 'claude_3' tests in the file? 

Thanks!

**Comment by nkvch** [comment]
@krrishdholakia 
it is okay now:
```
PS C:\Users\nkvch\projects\open_source\litellm> python -m pytest litellm\tests\test_completion.py -k claude_3
======================================== test session starts =========================================
platform win32 -- Python 3.11.4, pytest-8.2.0, pluggy-1.5.0
rootdir: C:\Users\nkvch\projects\open_source\litellm
configfile: pyproject.toml
plugins: anyio-4.2.0
collected 101 items / 93 deselected / 8 selected

litellm\tests\test_completion.py final cost: 0.008879999999999999; prompt_tokens_cost_usd_dollar: 0.00138; completion_tokens_cost_usd_dollar: 0.0075
success callbacks: []
.s...s..                                                       [100%]

========================================== warnings summary ========================================== 
venv\Lib\site-packages\pydantic\_internal\_config.py:284: 24 warnings
  C:\Users\nkvch\projects\open_source\litellm\venv\Lib\site-packages\pydantic\_internal\_config.py:284: PydanticDeprecatedSince20: Support for class-based `config` is deprecated, use ConfigDict instead. Deprecated in Pydantic V2.0 to be removed in V3.0. See Pydantic V2 Migration Guide at https://errors.pydantic.dev/2.7/migration/
    warnings.warn(DEPRECATION_MESSAGE, DeprecationWarning)

venv\Lib\site-packages\pydantic\_internal\_fields.py:160
  C:\Users\nkvch\projects\open_source\litellm\venv\Lib\site-packages\pydantic\_internal\_fields.py:160: UserWarning: Field "model_group_retry_policy" has conflict with protected namespace "model_".

  You may be able to resolve this warning by setting `model_config['protected_namespaces'] = ()`.      
    warnings.warn(

litellm\proxy\_types.py:219
  C:\Users\nkvch\projects\open_source\litellm\litellm\proxy\_types.py:219: PydanticDeprecatedSince20: Pydantic V1 style `@root_validator` validators are deprecated. You should migrate to Pydantic V2 style `@model_validator` validators, see the migration guide for more details. Deprecated in Pydantic V2.0 to be removed in V3.0. See Pydantic V2 Migration Guide at https://errors.pydantic.dev/2.7/migration/      
    @root_validator(pre=True)

litellm\proxy\_types.py:306
  C:\Users\nkvch\projects\open_source\litellm\litellm\proxy\_types.py:306: PydanticDeprecatedSince20: `pydantic.config.Extra` is deprecated, use literal values instead (e.g. `extra='allow'`). Deprecated in Pydantic V2.0 to be removed in V3.0. See Pydantic V2 Migration Guide at https://errors.pydantic.dev/2.7/migration/
    extra = Extra.allow  # Allow extra fields

litellm\proxy\_types.py:309
  C:\Users\nkvch\projects\open_source\litellm\litellm\proxy\_types.py:309: PydanticDeprecatedSince20: Pydantic V1 style `@root_validator` validators are deprecated. You should migrate to Pydantic V2 style `@model_validator` validators, see the migration guide for more details. Deprecated in Pydantic V2.0 to be removed in V3.0. See Pydantic V2 Migration Guide at https://errors.pydantic.dev/2.7/migration/      
    @root_validator(pre=True)

litellm\proxy\_types.py:338
  C:\Users\nkvch\projects\open_source\litellm\litellm\proxy\_types.py:338: PydanticDeprecatedSince20: Pydantic V1 style `@root_validator` validators are deprecated. You should migrate to Pydantic V2 style `@model_validator` validators, see the migration guide for more details. Deprecated in Pydantic V2.0 to be removed in V3.0. See Pydantic V2 Migration Guide at https://errors.pydantic.dev/2.7/migration/      
    @root_validator(pre=True)

litellm\proxy\_types.py:385
  C:\Users\nkvch\projects\open_source\litellm\litellm\proxy\_types.py:385: PydanticDeprecatedSince20: Pydantic V1 style `@root_validator` validators are deprecated. You should migrate to Pydantic V2 style `@model_validator` validators, see the migration guide for more details. Deprecated in Pydantic V2.0 to be removed in V3.0. See Pydantic V2 Migration Guide at https://errors.pydantic.dev/2.7/migration/      
    @root_validator(pre=True)

litellm\proxy\_types.py:454
  C:\Users\nkvch\projects\open_source\litellm\litellm\proxy\_types.py:454: PydanticDeprecatedSince20: Pydantic V1 style `@root_validator` validators are deprecated. You should migrate to Pydantic V2 style `@model_validator` validators, see the migration guide for more details. Deprecated in Pydantic V2.0 to be removed in V3.0. See Pydantic V2 Migration Guide at https://errors.pydantic.dev/2.7/migration/      
    @root_validator(pre=True)

litellm\proxy\_types.py:466
  C:\Users\nkvch\projects\open_source\litellm\litellm\proxy\_types.py:466: PydanticDeprecatedSince20: Pydantic V1 style `@root_validator` validators are deprecated. You should migrate to Pydantic V2 style `@model_validator` validators, see the migration guide for more details. Deprecated in Pydantic V2.0 to be removed in V3.0. See Pydantic V2 Migration Guide at https://errors.pydantic.dev/2.7/migration/      
    @root_validator(pre=True)

litellm\proxy\_types.py:509
  C:\Users\nkvch\projects\open_source\litellm\litellm\proxy\_types.py:509: PydanticDeprecatedSince20: Pydantic V1 style `@root_validator` validators are deprecated. You should migrate to Pydantic V2 style `@model_validator` validators, see the migration guide for more details. Deprecated in Pydantic V2.0 to be removed in V3.0. See Pydantic V2 Migration Guide at https://errors.pydantic.dev/2.7/migration/      
    @root_validator(pre=True)

litellm\proxy\_types.py:546
  C:\Users\nkvch\projects\open_source\litellm\litellm\proxy\_types.py:546: PydanticDeprecatedSince20: Pydantic V1 style `@root_validator` validators are deprecated. You should migrate to Pydantic V2 style `@model_validator` validators, see the migration guide for more details. Deprecated in Pydantic V2.0 to be removed in V3.0. See Pydantic V2 Migration Guide at https://errors.pydantic.dev/2.7/migration/      
    @root_validator(pre=True)

litellm\proxy\_types.py:840
  C:\Users\nkvch\projects\open_source\litellm\litellm\proxy\_types.py:840: PydanticDeprecatedSince20: Pydantic V1 style `@root_validator` validators are deprecated. You should migrate to Pydantic V2 style `@model_validator` validators, see the migration guide for more details. Deprecated in Pydantic V2.0 to be removed in V3.0. See Pydantic V2 Migration Guide at https://errors.pydantic.dev/2.7/migration/      
    @root_validator(pre=True)

litellm\proxy\_types.py:867
  C:\Users\nkvch\projects\open_source\litellm\litellm\proxy\_types.py:867: PydanticDeprecatedSince20: Pydantic V1 style `@root_validator` validators are deprecated. You should migrate to Pydantic V2 style `@model_validator` validators, see the migration guide for more details. Deprecated in Pydantic V2.0 to be removed in V3.0. See Pydantic V2 Migration Guide at https://errors.pydantic.dev/2.7/migration/      
    @root_validator(pre=True)

litellm\proxy\_types.py:886
  C:\Users\nkvch\projects\open_source\litellm\litellm\proxy\_types.py:886: PydanticDeprecatedSince20: Pydantic V1 style `@root_validator` validators are deprecated. You should migrate to Pydantic V2 style `@model_validator` validators, see the migration guide for more details. Deprecated in Pydantic V2.0 to be removed in V3.0. See Pydantic V2 Migration Guide at https://errors.pydantic.dev/2.7/migration/      
    @root_validator(pre=True)

litellm\tests\test_completion.py:267
  C:\Users\nkvch\projects\open_source\litellm\litellm\tests\test_completion.py:267: PytestUnknownMarkWarning: Unknown pytest.mark.asyncio - is this a typo?  You can register custom marks to avoid this warning - for details, see https://docs.pytest.org/en/stable/how-to/mark.html
    @pytest.mark.asyncio

litellm\tests\test_completion.py:825
  C:\Users\nkvch\projects\open_source\litellm\litellm\tests\test_completion.py:825: PytestUnknownMarkWarning: Unknown pytest.mark.asyncio - is this a typo?  You can register custom marks to avoid this warning - for details, see https://docs.pytest.org/en/stable/how-to/mark.html
    @pytest.mark.asyncio

litellm\tests\test_completion.py:1332
  C:\Users\nkvch\projects\open_source\litellm\litellm\tests\test_completion.py:1332: PytestUnknownMarkWarning: Unknown pytest.mark.asyncio - is this a typo?  You can register custom marks to avoid this warning - for details, see https://docs.pytest.org/en/stable/how-to/mark.html
    @pytest.mark.asyncio

litellm\tests\test_completion.py:1664
  C:\Users\nkvch\projects\open_source\litellm\litellm\tests\test_completion.py:1664: PytestUnknownMarkWarning: Unknown pytest.mark.asyncio - is this a typo?  You can register custom marks to avoid this warning - for details, see https://docs.pytest.org/en/stable/how-to/mark.html
    @pytest.mark.asyncio

litellm\tests\test_completion.py:1713
  C:\Users\nkvch\projects\open_source\litellm\litellm\tests\test_completion.py:1713: PytestUnknownMarkWarning: Unknown pytest.mark.asyncio - is this a typo?  You can register custom marks to avoid this warning - for details, see https://docs.pytest.org/en/stable/how-to/mark.html
    @pytest.mark.asyncio

litellm\tests\test_completion.py:1867
  C:\Users\nkvch\projects\open_source\litellm\litellm\tests\test_completion.py:1867: PytestUnknownMarkWarning: Unknown pytest.mark.asyncio - is this a typo?  You can register custom marks to avoid this warning - for details, see https://docs.pytest.org/en/stable/how-to/mark.html
    @pytest.mark.asyncio

litellm\tests\test_completion.py:2399
  C:\Users\nkvch\projects\open_source\litellm\litellm\tests\test_completion.py:2399: PytestUnknownMarkWarning: Unknown pytest.mark.asyncio - is this a typo?  You can register custom marks to avoid this warning - for details, see https://docs.pytest.org/en/stable/how-to/mark.html
    @pytest.mark.asyncio

litellm\tests\test_completion.py:3031
  C:\Users\nkvch\projects\open_source\litellm\litellm\tests\test_completion.py:3031: PytestUnknownMarkWarning: Unknown pytest.mark.asyncio - is this a typo?  You can register custom marks to avoid this warning - for details, see https://docs.pytest.org/en/stable/how-to/mark.html
    @pytest.mark.asyncio

litellm\tests\test_completion.py:3154
  C:\Users\nkvch\projects\open_source\litellm\litellm\tests\test_completion.py:3154: PytestUnknownMarkWarning: Unknown pytest.mark.asyncio - is this a typo?  You can register custom marks to avoid this warning - for details, see https://docs.pytest.org/en/stable/how-to/mark.html
    @pytest.mark.asyncio

litellm/tests/test_completion.py::test_completion_claude_3_stream
litellm/tests/test_completion.py::test_completion_claude_3_stream
litellm/tests/test_completion.py::test_completion_claude_3_stream
litellm\utils.py:3776
  C:\Users\nkvch\projects\open_source\litellm\litellm\utils.py:3776: DeprecationWarning: open_text is deprecated. Use files() instead. Refer to https://importlib-resources.readthedocs.io/en/latest/using.html#migrating-from-legacy for migration advice.
    with resources.open_text(

-- Docs: https://docs.pytest.org/en/stable/how-to/capture-warnings.html
===================== 6 passed, 2 skipped, 93 deselected, 50 warnings in 58.56s ====================== 
PS C:\Users\nkvch\projects\open_source\litellm>
```

File: litellm/llms/prompt_templates/factory.py [hunk 1]
```diff
@@ -1016,7 +1014,13 @@ def anthropic_messages_pt(messages: list):
             msg_i += 1
 
         if user_content:
-            new_messages.append({"role": "user", "content": user_content})
+            if merge_with_previous:
+                new_messages[-1]["content"].extend(user_content)
```
Comment (ghost) [inline litellm/llms/prompt_templates/factory.py:1018]: this assumes new_messages will always contain "content" if 'merge_with_previous' is true 

however -> merge_with_previous seems to be set if msg_i > 0, not if new_messages contains at least one item of content

**Comment by ghost** [comment]
@nkvch had to revert the initial PR due to this concern ^ i can take some more time to review tomorrow. Let me know if my initial read of the issue is correct

**Pull request merged**